plugins/database/src/db_plugin.cpp

```diff
@@ -9197,95 +9263,71 @@ irods::error db_add_avu_metadata_op(
         }
-        if ( status < 0 ) {
-            _rollback( "chlAddAVUMetadata" );
-            return ERROR( status, "select data_id failed" );
+
+        status = cmlCheckDataObjOnly(logicalParentDirName, logicalEndName,
+                                     _ctx.comm()->clientUser.userName,
+                                     _ctx.comm()->clientUser.rodsZone,
+                                     ACCESS_CREATE_METADATA, &icss, admin_mode);
+
+        if (status < 0) {
+            _rollback("chlAddAVUMetadata");
+            return ERROR(status, "select data_id failed");
         }
+
         objId = status;
     }
 
-    if ( itype == 2 ) {
-        if ( _admin_mode == 1 ) {
-            if ( logSQL != 0 ) {
-                rodsLog( LOG_SQL, "chlAddAVUMetadata SQL 3" );
-            }
-            {
-                std::vector<std::string> bindVars;
-                bindVars.push_back( _name );
-                status = cmlGetIntegerValueFromSql(
-                             "select coll_id from R_COLL_MAIN where coll_name=?",
-                             &iVal, bindVars, &icss );
-            }
-            if ( status == 0 ) {
-                status = iVal;    /*like cmlCheckDir, status is objid*/
-            }
-        }
-        else {
-            /* Check that the collection exists and user has create_metadata
-               permission, and get the collectionID */
-            if ( logSQL != 0 ) {
-                rodsLog( LOG_SQL, "chlAddAVUMetadata SQL 4" );
-            }
-            status = cmlCheckDir( _name,
-                                  _ctx.comm()->clientUser.userName,
-                                  _ctx.comm()->clientUser.rodsZone,
-                                  ACCESS_CREATE_METADATA, &icss );
+    // Collections
+    if (itype == 2) {
+        // Check that the collection exists and user has create_metadata
+        // permission, and get the collectionID.
+        if ( logSQL != 0 ) {
+            rodsLog( LOG_SQL, "chlAddAVUMetadata SQL 4" );
         }
+
+        status = cmlCheckDir(_name,
+                             _ctx.comm()->clientUser.userName,
+                             _ctx.comm()->clientUser.rodsZone,
+                             ACCESS_CREATE_METADATA, &icss, admin_mode);
+
         if ( status < 0 ) {
             char errMsg[105];
-            _rollback( "chlAddAVUMetadata" );
+
+            _rollback( "chlAddAVUMetadata" ); // TODO We've rolled back here, so why the extra call below?
+
             if ( status == CAT_UNKNOWN_COLLECTION ) {
-                snprintf( errMsg, 100, "collection '%s' is unknown",
-                          _name );
+                snprintf( errMsg, 100, "collection '%s' is unknown", _name );
                 addRErrorMsg( &_ctx.comm()->rError, 0, errMsg );
             }
             else {
-                _rollback( "chlAddAVUMetadata" );
+                _rollback( "chlAddAVUMetadata" ); // TODO Why do we rollback again?
```

Comment:
investigate TODO comments or capture in an issue for later.

this PR does not need to handle this.

Reply:
done